### PR TITLE
Added the small batch on the gdplib

### PR DESCRIPTION
**Pull request by AlbertLee125**
Added the gdp_small_batch.py on the gdp library.

### CODE REVIEW

File: gdplib/small_batch/gdp_small_batch.py [hunk 1]
```diff
@@ -0,0 +1,460 @@
...
+        m,
+        tee=True,
+        # Uncomment the following lines if you want to save GAMS models
+        # keepfiles=True,
+        # tmpdir=gams_path,
+        # symbolic_solver_labels=True,
+        add_options=[
+            'option reslim = ' + str(timelimit) + ';'
+            'option optcr = 0.0;'
+            # Uncomment the following lines to setup IIS computation of BARON through option file
```
Comment (ZedongPeng) [inline gdplib/small_batch/gdp_small_batch.py]: I think it is better for us to create an option to control if we want to set up the IIS computation.

File: gdplib/small_batch/gdp_small_batch.py [hunk 2]
```diff
@@ -0,0 +1,460 @@
...
+    """
+    # Transformation step
+    pe.TransformationFactory('core.logical_to_linear').apply_to(m)
+    transformation_string = 'gdp.' + transformation
+    pe.TransformationFactory(transformation_string).apply_to(m)
+
+    # Solution step
+    dir_path = os.path.dirname(os.path.abspath(__file__))
+    gams_path = os.path.join(dir_path, "gamsfiles/")
+    if not (os.path.exists(gams_path)):
```
Comment (ZedongPeng) [inline gdplib/small_batch/gdp_small_batch.py]: Why do we need to check the `gamsfiles` here?

File: gdplib/small_batch/gdp_small_batch.py [hunk 3]
```diff
@@ -0,0 +1,460 @@
...
+        """
+        return sum(m.alpha[j] * (pe.exp(m.n[j] + m.beta[j] * m.v[j])) for j in m.j)
+
+    m.obj = pe.Objective(rule=obj_rule, sense=pe.minimize)
+
+    return m
+
+
+def external_ref(m, x, logic_expr=None):
+    """
```
Comment (ZedongPeng) [inline gdplib/small_batch/gdp_small_batch.py]: If I understand correctly, this function aims to fix the disjunctions according to the provided variables and return the fixed model. We can come up with a better name for this function.

File: gdplib/small_batch/gdp_small_batch.py [hunk 4]
```diff
@@ -0,0 +1,460 @@
...
+
+    # EXTERNAL REF TEST (this thest can be deleted)
+    newmodel = external_ref(m, [1, 2, 3], logic_expr=None)
+    # print('External Ref Test')
+    # print('Y[1, mixer] = ', newmodel.Y[1, 'mixer'].value)
+    # print('Y_exists[1, mixer] = ', newmodel.Y_exists[1, 'mixer'].indicator_var.value)
+    # print('Y_not_exists[1, mixer] = ', newmodel.Y_not_exists[1, 'mixer'].indicator_var.value)
+    # print('Y[2, mixer] = ', newmodel.Y[2, 'mixer'].value)
+    # print('Y_exists[2, mixer] = ', newmodel.Y_exists[2, 'mixer'].indicator_var.value)
+    # print('Y_not_exists[2, mixer] = ', newmodel.Y_not_exists[2, 'mixer'].indicator_var.value)
```
Comment (ZedongPeng) [inline gdplib/small_batch/gdp_small_batch.py]: If these lines are only used for debugging, we can remove them.

**Review by ZedongPeng — CHANGES_REQUESTED** [review]
Good and well-structured PR.

**Comment by bernalde** [comment]
I will only want us to work on this PR once all the other examples from GDPLib are updated to the latest version of Pyomo.

File: gdplib/small_batch/gdp_small_batch.py [hunk 5]
```diff
@@ -0,0 +1,460 @@
+"""
+gdp_small_batch.py
+The gdp_small_batch.py module contains the GDP model for the small batch problem based on the Kocis and Grossmann (1988) paper.
+The problem is based on the Example 4 of the paper.
+
+References:
+    - Kocis, G. R.; Grossmann, I. E. Global Optimization of Nonconvex Mixed-Integer Nonlinear Programming (MINLP) Problems in Process Synthesis. Ind. Eng. Chem. Res. 1988, 27 (8), 1407–1421.
```
Comment (bernalde) [inline gdplib/small_batch/gdp_small_batch.py]: DOI missing

File: gdplib/small_batch/gdp_small_batch.py [hunk 6]
```diff
@@ -0,0 +1,460 @@
...
+)
+from pyomo.gdp import Disjunct, Disjunction
+from pyomo.opt.base.solvers import SolverFactory
+
+
+def build_small_batch():
+    """
+    The function build the GDP model for the small batch problem.
+
+    References:
```
Comment (bernalde) [inline gdplib/small_batch/gdp_small_batch.py]: DOI

File: gdplib/small_batch/gdp_small_batch.py (hunk 3 above)
Comment (bernalde) [reply gdplib/small_batch/gdp_small_batch.py]: I agree with Zedong. This should work not only for the LDSDA but in general. Call it fix disjunctions. ALso, do we really need this now that there is the new implementation of LDSDA? @ZedongPeng

File: gdplib/small_batch/gdp_small_batch.py [hunk 7]
```diff
@@ -0,0 +1,460 @@
...
+    p = 0
+    for j in m.j:
+        ext_var[j] = x[p]
+        p = p + 1
+
+    for k in m.k:
+        for j in m.j:
+            if k == ext_var[j]:
+                m.Y[k, j].fix(True)
+                # m.Y_exists[k, j].indicator_var.fix(
```
Comment (bernalde) [inline gdplib/small_batch/gdp_small_batch.py]: These comments can go

File: gdplib/small_batch/gdp_small_batch.py (hunk 2 above)
Comment (bernalde) [reply gdplib/small_batch/gdp_small_batch.py]: This function altogether is no longer needed. This was for the older implementation of LDSDA. Please remove it

File: gdplib/small_batch/gdp_small_batch.py (hunk 4 above)
Comment (bernalde) [reply gdplib/small_batch/gdp_small_batch.py]: I agree, we need to clean this up

**Review by bernalde — CHANGES_REQUESTED** [review]
Since this is the first PR with our own newly contributed models, we need to be way more strict with the formatting. Debugging comments are no longer allowed. Only functions relevant to the model should be included (not the solution method). THe example itself is not the one found in Kocis and Grossmann, it is a simplification made for the LDSDA paper, so update the references.

File: gdplib/small_batch/gdp_small_batch.py (hunk 6 above)
Comment (AlbertLee125) [reply gdplib/small_batch/gdp_small_batch.py]: DOI added and added the LD-SDA paper.

File: gdplib/small_batch/gdp_small_batch.py (hunk 5 above)
Comment (AlbertLee125) [reply gdplib/small_batch/gdp_small_batch.py]: DOI added and added the LD-SDA paper.

File: gdplib/small_batch/gdp_small_batch.py (hunk 2 above)
Comment (AlbertLee125) [reply gdplib/small_batch/gdp_small_batch.py]: I will remove this.

File: gdplib/small_batch/gdp_small_batch.py (hunk 1 above)
Comment (AlbertLee125) [reply gdplib/small_batch/gdp_small_batch.py]: This part will be removed.

File: gdplib/small_batch/gdp_small_batch.py (hunk 7 above)
Comment (AlbertLee125) [reply gdplib/small_batch/gdp_small_batch.py]: Comments removed.

File: gdplib/small_batch/gdp_small_batch.py (hunk 4 above)
Comment (AlbertLee125) [reply gdplib/small_batch/gdp_small_batch.py]: cleaned up the external ref test.

File: gdplib/small_batch/gdp_small_batch.py (hunk 3 above)
Comment (AlbertLee125) [reply gdplib/small_batch/gdp_small_batch.py]: Zedong's new LDSDA does not require these anymore. Therefore, I just removed it since we need the model.

**Review by AlbertLee125 — COMMENTED** [review]
I renewed the documentation of the small_batch.py and fixed it from Google format to Numpy format.

**Review by bernalde — APPROVED** [review]
This looks good, @ZedongPeng if this code already runs with your LDSDA implementation, feel free to merge

**Comment by bernalde** [comment]
Replace gam by gamma as it is the ration of Cp and Cv

**Comment by AlbertLee125** [comment]
The spelling check is failing because of the hda files. Besides the spelling check, there are no issues with the gdp_small_batch.py and I strongly believe this is ready to be merged. And I am not sure whether we have to discuss for the format of HDA files, There are two gammas one is the gamma for the ratio of the Cv and Cp the other is the decision variables related to the absorption process.

File: gdplib/small_batch/README.md [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+
+The objective is to minimize the investment cost of the batch units.
+
+The solution is 167427.65711.
+
+### References
+
+[1] Kocis, G. R.; Grossmann, I. E. Global Optimization of Nonconvex Mixed-Integer Nonlinear Programming (MINLP) Problems in Process Synthesis. Ind. Eng. Chem. Res. 1988, 27 (8), 1407-1421. https://doi.org/10.1021/ie00080a013
+
+[2] Ovalle, D., Liñán, D. A., Lee, A., Gómez, J. M., Ricardez-Sandoval, L., Grossmann, I. E., & Neira, D. E. B. (2024). Logic-Based Discrete-Steepest Descent: A Solution Method for Process Synthesis Generalized Disjunctive Programs. arXiv preprint arXiv:2405.05358. https://doi.org/10.48550/arXiv.2405.05358
```
Comment (bernalde) [inline gdplib/small_batch/README.md]: My last name to be cited is Bernal Neira, not Neira

File: gdplib/small_batch/gdp_small_batch.py [hunk 9]
```diff
@@ -9,6 +9,7 @@
 [1] Kocis, G. R.; Grossmann, I. E. Global Optimization of Nonconvex Mixed-Integer Nonlinear Programming (MINLP) Problems in Process Synthesis. Ind. Eng. Chem. Res. 1988, 27 (8), 1407-1421. https://doi.org/10.1021/ie00080a013
 [2] Ovalle, D., Liñán, D. A., Lee, A., Gómez, J. M., Ricardez-Sandoval, L., Grossmann, I. E., & Neira, D. E. B. (2024). Logic-Based Discrete-Steepest Descent: A Solution Method for Process Synthesis Generalized Disjunctive Programs. arXiv preprint arXiv:2405.05358. https://doi.org/10.48550/arXiv.2405.05358
```
Comment (bernalde) [inline gdplib/small_batch/gdp_small_batch.py]: See previous comment regarding last name. Fix elsewhere.

**Review by bernalde — CHANGES_REQUESTED** [review]
There is a typo is my last name in the references and the master branch needs to be merged before the tests will pass.

File: gdplib/small_batch/gdp_small_batch.py (hunk 9 above)
Comment (AlbertLee125) [reply gdplib/small_batch/gdp_small_batch.py]: Replaced Neira, D. E. B. into Bernal Neira, D. E.

**Review by bernalde — APPROVED** [review]

**Review by ZedongPeng — APPROVED** [review]

**Pull request merged**